### PR TITLE
End statements correctly

### PR DESCRIPTION
**Pull request by EMattfolk**
Tests still need to be written...

### CODE REVIEW

File: sylt-parser/src/expression.rs [hunk 1]
```diff
@@ -169,15 +169,7 @@ fn function<'t>(ctx: Context<'t>) -> ParseResult<'t, Expression> {
     };
 
     // Parse the function statement.
-    let (ctx, mut statement) = block_statement(ctx)?;
-
-    // If the return type isn't void, check for and apply implicit returns.
```
Comment (FredTheDino) [inline sylt-parser/src/expression.rs:174]: Keep this.

File: sylt-parser/src/expression.rs [hunk 2]
```diff
@@ -169,15 +169,7 @@ fn function<'t>(ctx: Context<'t>) -> ParseResult<'t, Expression> {
...
-
-    let statements = if let StatementKind::Block { statements } = &mut statement.kind {
-        statements
-    } else {
-        unreachable!("Function blocks should only be blocks");
-    };
+    let (ctx, mut statements) = block(ctx)?;
 
     if !matches!(ret.kind, Resolved(Void)) {
```
Comment (FredTheDino) [inline sylt-parser/src/expression.rs:174]:
```suggestion
    // If the return type isn't void, check for and apply implicit returns.
    if !matches!(ret.kind, Resolved(Void)) {
```

File: sylt-parser/src/parser.rs [hunk 3]
```diff
@@ -219,6 +219,13 @@ impl<'a> Context<'a> {
         new
     }
 
+    /// Back up one token. Will not move past the beginning.
```
Comment (FredTheDino) [inline sylt-parser/src/parser.rs:222]:
```suggestion
    /// Back up one token. Will not move before the beginning.
```
Is it clearer or nah?

File: sylt-parser/src/statement.rs [hunk 4]
```diff
@@ -120,8 +120,7 @@ pub struct Statement {
     pub kind: StatementKind,
 }
 
-pub fn block_statement<'t>(ctx: Context<'t>) -> ParseResult<'t, Statement> {
-    let span = ctx.span();
+pub fn block<'t>(ctx: Context<'t>) -> ParseResult<'t, Vec<Statement>> {
```
Comment (FredTheDino) [inline sylt-parser/src/statement.rs:123]: +1

File: sylt-parser/src/statement.rs [hunk 5]
```diff
@@ -144,10 +143,10 @@ pub fn block_statement<'t>(ctx: Context<'t>) -> ParseResult<'t, Statement> {
         }
     }
 
-    let ctx = expect!(ctx, T::RightBrace, "Expected }} after block statement");
     if errs.is_empty() {
+        let ctx = expect!(ctx, T::RightBrace, "Expected }} after block statement");
```
Comment (FredTheDino) [inline sylt-parser/src/statement.rs:147]:
```suggestion
        let ctx = expect!(ctx, T::RightBrace, "Expected }} after block");
```

File: sylt-parser/src/statement.rs [hunk 6]
```diff
@@ -457,7 +456,13 @@ pub fn statement<'t>(ctx: Context<'t>) -> ParseResult<'t, Statement> {
...
     };
 
-    let ctx = ctx.skip_if(T::Newline);
+    // Newline, RightBrace and Else can end a statment.
+    // We don't give too specific errors when that is not the case.
+    let ctx = if !matches!(ctx.token(), T::RightBrace | T::Else) {
+        expect!(ctx, T::Newline, "Expected newline")
+    } else {
+        ctx
+    };
```
Comment (FredTheDino) [inline sylt-parser/src/statement.rs:465]:
```suggestion
    let ctx = if matches!(ctx.token(), T::RightBrace | T::Else) {
		ctx
    } else {
        expect!(ctx, T::Newline, "Expected newline to end statement")
    };
```

File: sylt-parser/src/statement.rs [hunk 7]
```diff
@@ -457,7 +456,13 @@ pub fn statement<'t>(ctx: Context<'t>) -> ParseResult<'t, Statement> {
         }
     };
 
-    let ctx = ctx.skip_if(T::Newline);
+    // Newline, RightBrace and Else can end a statment.
+    // We don't give too specific errors when that is not the case.
```
Comment (FredTheDino) [inline sylt-parser/src/statement.rs:460]:
```suggestion
    // Errors only report on the newline - since it's usually the solution.
```
Maybe? I dislike the double negation in the comment - makes it hard to follow.

**Review by FredTheDino — APPROVED** [review]
Nice! Great work! <3

File: sylt-parser/src/statement.rs (hunk 6 above)
Comment (sornas) [reply sylt-parser/src/statement.rs:465]: This suggestion has tabs!!

File: sylt-parser/src/statement.rs (hunk 7 above)
Comment (sornas) [reply sylt-parser/src/statement.rs:460]: What's "the solution"?

File: sylt-parser/src/parser.rs (hunk 3 above)
Comment (EMattfolk) [reply sylt-parser/src/parser.rs:222]: I would say nah

File: sylt-parser/src/statement.rs (hunk 7 above)
Comment (EMattfolk) [reply sylt-parser/src/statement.rs:460]:
```suggestion
    // If a statement does not end, we only report it as a missing newline.
```
Perhaps?
